--- gcode.py.orig	2019-07-07 14:10:03.000000000 +0300
+++ gcode.py	2019-07-14 16:06:37.432801986 +0300
@@ -27,6 +27,7 @@
                                                       self._process_data)
         self.partial_input = ""
         self.pending_commands = []
+	self.respond_callbacks = []
         self.bytes_read = 0
         self.input_log = collections.deque([], 50)
         # Command handling
@@ -304,11 +305,15 @@
         except os.error:
             logging.exception("Write g-code ack")
         self.need_ack = False
+    def register_respond_callback(self, callback):
+        self.respond_callbacks.append(callback)
     def respond(self, msg):
         if self.is_fileinput:
             return
         try:
             os.write(self.fd, msg+"\n")
+            for callback in self.respond_callbacks:
+                callback(msg+"\n")
         except os.error:
             logging.exception("Write g-code response")
     def respond_info(self, msg, log=True):
